vllm/entrypoints/chat_utils.py

```diff
@@ -143,13 +194,37 @@ def _mm_token_str(model_config: ModelConfig, tokenizer: AnyTokenizer,
 
 # TODO: Let user specify how to insert multimodal tokens into prompt
 # (similar to chat template)
-def _get_full_multimodal_text_prompt(placeholder_token_str: str,
+def _get_full_multimodal_text_prompt(placeholders: List[str],
                                      text_prompt: str) -> str:
     """Combine multimodal prompts for a multimodal language model"""
 
+    # Look through the text prompt so that we don't add placeholders for
+    # items that already have them.
+    prompt_fragments = [text_prompt]
+    missing_placeholders: List[str] = []
+    for placeholder in placeholders:
```

Comment:
This is pretty convoluted as it's a full generalization of the check that was there before -- happy to do something much simpler, I just wasn't sure which "placeholder is already present in the prompt" scenarios are important.

Reply:
(Also it needs tests if it remains in it's current form :))

Reply:
I think this is fine, but definitely needs some examples to better make sense of it.